src/core/reactor_backend.cc

```diff
@@ -1484,10 +1490,97 @@ class reactor_backend_uring final : public reactor_backend {
         delete pfd;
     }
     virtual future<std::tuple<pollable_fd, socket_address>> accept(pollable_fd_state& listenfd) override {
-        return _r.do_accept(listenfd);
+        if (listenfd.take_speculation(POLLIN | POLLOUT)) {
```

Comment:
I don't think we need POLLOUT any more, since we don't poll the socket at all. See c879bb52f764bf8899c8f45c9240d969c0994094 why it was introduced.

It doesn't hurt, but it can puzzle readers.

Reply:
And please make sure the tests include shutdown on a listening socket. I'm guessing httpd+ctrl-C is enough, but please verify.

Reply:
> I don't think we need POLLOUT any more, since we don't poll the socket at all. See [c879bb5](https://github.com/scylladb/seastar/commit/c879bb52f764bf8899c8f45c9240d969c0994094) why it was introduced.
> 
> It doesn't hurt, but it can puzzle readers.

see #1248. and unfortunately, `accept4()` does not allow us to pass `MSG_DONTWAIT` to it to consume the read buffer without being blocked if the listenfd is marked without `SOCK_NONBLOCK`. so, in order to do speculated `accept`,

1. we still need to open the listenfd with `SOCK_NONBLOCK`
2. if there is no data in the input buffer of `listenfd`, cqe of the `accept4()` op returns `-EAGAIN`.
3. in order to avoid "busy polling" `accept()` ops returning `-EAGAIN`, we have to wait for `EPOLLIN` or `POLLOUT`. before sending another `accept4()` to the ring.

Reply:
> And please make sure the tests include shutdown on a listening socket. I'm guessing httpd+ctrl-C is enough, but please verify.

sure. tested with following steps:

1. run `httpd --reactor-backend io_uring --smp 1`
2. send ctrl-c in the same terminal

and 

1. run `httpd --reactor-backend io_uring --smp 1`
2. run `wrk -c 128 -t 4 http://0.0.0.0:10000`
3. while `wrk` is still running, send ctrl-c in the same terminal of httpd

in both test cases, the httpd server stopped without issues or errors.

Reply:
> > I don't think we need POLLOUT any more, since we don't poll the socket at all. See [c879bb5](https://github.com/scylladb/seastar/commit/c879bb52f764bf8899c8f45c9240d969c0994094) why it was introduced.
> > It doesn't hurt, but it can puzzle readers.
> 
> see #1248. and unfortunately, `accept4()` does not allow us to pass `MSG_DONTWAIT` to it to consume the read buffer without being blocked if the listenfd is marked without `SOCK_NONBLOCK`. so, in order to do speculated `accept`,
> 
> 1. we still need to open the listenfd with `SOCK_NONBLOCK`
> 2. if there is no data in the input buffer of `listenfd`, cqe of the `accept4()` op returns `-EAGAIN`.
> 3. in order to avoid "busy polling" `accept()` ops returning `-EAGAIN`, we have to wait for `EPOLLIN` or `POLLOUT`. before sending another `accept4()` to the ring.

I agree we should continue to use SOCK_NONBLOCK, but why do we have to check POLLIN|POLLOUT? the uring accept doesn't go through uring poll, so pollable_fd_state::events_known is only ever updated by speculation. Since we only write POLLIN, we'll never see POLLOUT.

Reply:
> but why do we have to check POLLIN|POLLOUT? the uring accept doesn't go through uring poll, so pollable_fd_state::events_known is only ever updated by speculation. Since we only write POLLIN, we'll never see POLLOUT.

thank you, Avi! right! sorry i misunderstood your point. i was referencing https://github.com/scylladb/seastar/commit/c879bb52f764bf8899c8f45c9240d969c0994094

> A complication is that after shutdown() on a listening socket, epoll_wait() return with EPOLLOUT|EPOLLHUP

so i assumed that `reactor_backend_uring::uring_pollable_fd_state::get_desc()` would return a completion which sets the fd's `events_known` on completion. but i was wrong! if the fd is closed, an exception is thrown. that's what `_listenfd.maybe_no_more_recv()` is for.

so, yes. will drop it.